In [2]:
def welcome_message():
    print("hello")


In [3]:
welcome_message()

hello


In [6]:
def welcome_message(user):
    print(f"hello {user} ") #line1, line2 --set of instruction

welcome_message("Mr Fahim") #eta input
#output hosse result ta

hello Mr Fahim 


In [7]:
def welcome_message(user="none"):
    print(f"hello {user} ") #line1, line2 --set of instruction

welcome_message() #kono input na dile none dekabe
#output hosse result ta

hello none 


### 2.3 Structure of a func - Return and Doc

In [11]:
def adder(a,b):
    sum = a+b
    print(sum) #line 1-3 instruction

adder (10,30)

40


In [ ]:
def adder(a,b):
    """ 
    added two number
    """

    sum = a+b
    return sum #line 1-3 instruction

equation= adder (10,30) * 5 #equation solve korte hole return use korte hbe
print(equation)

#help(adder) #help diye function ta ki bujasse ta jana jabe """string msg diye"

200
Help on function adder in module __main__:

adder(a, b)
    added two number



### 2.4 Positional Arguments


In [ ]:
def adder(a,b):
    print(f"{a}, {b}")
    
adder(10,20)

10, 20


In [ ]:
def adder(*args):
    print(args) # jokon onk argument dite hbe *args use korbo
    
adder(10,20,5,51)

(10, 20, 5, 51)


In [20]:
#loop calano jabe tuple e

def adder(*args):
    for i in args:

         print(i) # jokon onk argument dite hbe *args use korbo
    
adder(10,20,5,51,23)

10
20
5
51
23


### 2.5 Keyword Arguments


In [ ]:
def adder(a,b,c):
    print(f"{a}, {b}, {c}")
    
adder(a=10, c=20, b=44) #keyword 

10, 44, 20


In [30]:
def adder(**kwargs):

    print(kwargs)

adder(a=10, c=20, b=44, n_estimator=50)

{'a': 10, 'c': 20, 'b': 44, 'n_estimator': 50}


In [31]:
def adder(**kwargs):
    for key,value in kwargs.items():



        print(f"{key}:{value}")

adder(a=10, c=20, b=44, n_estimator=50)

a:10
c:20
b:44
n_estimator:50


### 2.6 Return Type of a Function


In [33]:
def welcome_message():

    print("hello user") 

print(type(welcome_message()))


hello user
<class 'NoneType'>


In [4]:
def welcome_message(user="none"):


    return f"hello {user}" #print er bodle return use kora jabe

print(type(welcome_message()))

<class 'str'>


In [6]:
def adder(a,b):
    return a,b
     #line 1-3 instruction

print(type(adder(10.5, 8)))


#unzip

x,y = adder(10.5, 8)

print(y)

<class 'tuple'>
8


In [ ]:
def adder(a,b):
    return [a,b] #list o kora jai
     #line 1-3 instruction

print(type(adder(10.5, 8)))

<class 'list'>


### 2.7 Iterator and Generator

In [11]:
s = {10,20,120,33,44}

s_iter = iter(s)

print(type(s_iter))

<class 'set_iterator'>


In [17]:
s = {10,20,120,33,44}

s_iter = iter(s) #iter use kora hoi set k index er moto dekanor jnno

print(next(s_iter)) #next diye unorderly ektar por ekta indexing kora hoi
print(next(s_iter))
print(next(s_iter))
print(next(s_iter))
print(next(s_iter)) #unorderly

33
20
120
10
44


In [19]:
for i in s:
    print(i)

33
20
120
10
44


### generator 

In [ ]:
lst = [x for x in range(300)] #data set
print(lst)

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99, 100, 101, 102, 103, 104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116, 117, 118, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 130, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 144, 145, 146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158, 159, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 173, 174, 175, 176, 177, 178, 179, 180, 181, 182, 183, 184, 185, 186, 187, 188, 189, 190, 191, 192, 193, 194, 195, 196, 197, 198, 199, 200, 201, 202, 203, 204, 205, 206, 207, 208, 209, 210, 211, 212, 213, 214, 215, 216, 217, 218, 219, 220, 221,

In [ ]:
lst = [x for x in range(200)] #data set

def data_loader(chunk_size, lst):
    for i in range(len(lst)):
        yield lst [i:i+chunk_size]

x = data_loader(5, lst)
print(type(x))  


<class 'generator'>


In [27]:
lst = [x for x in range(200)] #data set

def data_loader(chunk_size, lst):
    for i in range(len(lst)):
        yield lst [i:i+chunk_size]

x = data_loader(5, lst)

print(x) 

<generator object data_loader at 0x00000202464273D0>


In [30]:
lst = [x for x in range(200)] #data set

def data_loader(chunk_size, lst):
    for i in range(len(lst)):
        yield lst [i:i+chunk_size]

x = data_loader(5, lst)
print(next(x))  
print(next(x))  
print(next(x))  

[0, 1, 2, 3, 4]
[1, 2, 3, 4, 5]
[2, 3, 4, 5, 6]


In [31]:
lst = [x for x in range(200)] #data set

def data_loader(chunk_size, lst):
    for i in range(0,len(lst), chunk_size):
        yield lst [i:i+chunk_size]

x = data_loader(5, lst)
print(next(x))  
print(next(x))  
print(next(x)) 

[0, 1, 2, 3, 4]
[5, 6, 7, 8, 9]
[10, 11, 12, 13, 14]


### 2.8 Lambda , Map and Apply function

### practice


### #problem 1
 Write a function introduce(name, age) that prints:

 My name is <name> and I am <age> years old.
 
Call it using positional arguments.


In [2]:
def introduce(name, age):
    print(f"My name is {name} and I am {age} years old.")
introduce("Rahim", 28)

My name is Rahim and I am 28 years old.


### problem 2
Write a Python generator function called batch_generator. It should take a list of numbers (representing data samples) and a batch_size. The generator should yield one batch at a time as a list.



In [3]:
def batch_generator(data, batch_size):
    for i in range(0, len(data), batch_size):
        yield data[i:i + batch_size]

# Usage
data = [1, 2, 3, 4, 5, 6, 7, 8]
batch_size = 3

for batch in batch_generator(data, batch_size):
    print(batch)

[1, 2, 3]
[4, 5, 6]
[7, 8]


In [ ]:
def batch_generator(data, batch_size):
    for i in range(0, len(data), batch_size):
        yield data[i:i + batch_size]

# Usage
data = [1, 2, 3, 4, 5, 6, 7, 8]
batch_size = 3

x = batch_generator(data, batch_size) #bikolpo
print(next(x))  
print(next(x))  
print(next(x))  


[1, 2, 3]
[4, 5, 6]
[7, 8]


### problem 3
Given a list [1,2,3,4], use map() and lambda to create a new list with squares of each number.

Sample Input:
 No input

Sample Output:
[1, 4, 9, 16]


In [8]:
numbers = [1, 2, 3, 4]
squares = list(map(lambda x: x**2, numbers))
print(squares)

[1, 4, 9, 16]


### problem 4
 Write a function power(base, exponent) that returns the result of base raised to exponent. If the exponent is not given, it should calculate the square.
Sample Input:
5
Sample Output:
25
Sample Input 2:
5 3
Sample Output 2:
125


In [10]:
def power(base, exponent=2):
    return base ** exponent

# Sample 1: only base given → squares by default
print(power(5))

# Sample 2: both base and exponent given
print(power(5, 3))

25
125


### problem 5
Given a list [1,2,3,4,5,6], use filter() to select even numbers and map() with lambda to square them.
Sample Input:
 No input
Sample Output:
[4, 16, 36]


In [9]:
numbers = [1, 2, 3, 4, 5, 6]
result = list(map(lambda x: x**2, filter(lambda x: x % 2 == 0, numbers)))
print(result)

[4, 16, 36]
